File: quasar-backend/src/controllers/speechController.js
const { pool } = require('../config/database');
const logger = require('../utils/logger');

/**
 * POST /api/sessions/:sessionId/speech-metrics
 * Save speech analysis data for a session
 */
async function saveSpeechMetrics(req, res) {
  try {
    const { sessionId } = req.params;
    const { transcript, fillerBuckets, totalFillers, wordsPerMinute } = req.body;

    if (!transcript && !fillerBuckets) {
      return res.status(400).json({ success: false, message: 'transcript or fillerBuckets required', data: null });
    }

    // Check session exists
    const sessionCheck = await pool.query('SELECT id FROM sessions WHERE id = $1', [sessionId]);
    if (sessionCheck.rows.length === 0) {
      return res.status(404).json({ success: false, message: 'Session not found', data: null });
    }

    // Upsert speech metrics
    const existing = await pool.query('SELECT id FROM speech_metrics WHERE session_id = $1', [sessionId]);

    let result;
    if (existing.rows.length > 0) {
      result = await pool.query(
        `UPDATE speech_metrics SET
          transcript = $1,
          filler_buckets = $2,
          total_fillers = $3,
          words_per_minute = $4
         WHERE session_id = $5
         RETURNING *`,
        [transcript || '', JSON.stringify(fillerBuckets || []), totalFillers || 0, wordsPerMinute || 0, sessionId]
      );
    } else {
      result = await pool.query(
        `INSERT INTO speech_metrics (session_id, transcript, filler_buckets, total_fillers, words_per_minute)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING *`,
        [sessionId, transcript || '', JSON.stringify(fillerBuckets || []), totalFillers || 0, wordsPerMinute || 0]
      );
    }

    return res.json({
      success: true,
      message: 'Speech metrics saved',
      data: result.rows[0]
    });
  } catch (err) {
    logger.error('Save speech metrics error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to save speech metrics', data: null });
  }
}

/**
 * GET /api/sessions/:sessionId/speech-metrics
 * Retrieve speech metrics for heatmap rendering
 */
async function getSpeechMetrics(req, res) {
  try {
    const { sessionId } = req.params;

    const result = await pool.query(
      'SELECT * FROM speech_metrics WHERE session_id = $1',
      [sessionId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ success: false, message: 'Speech metrics not found', data: null });
    }

    return res.json({
      success: true,
      message: 'Speech metrics retrieved',
      data: result.rows[0]
    });
  } catch (err) {
    logger.error('Get speech metrics error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to get speech metrics', data: null });
  }
}

module.exports = { saveSpeechMetrics, getSpeechMetrics };

File: backend/src/main/java/com/teamzemo/quasar/controller/SpeechController.java
package com.teamzemo.quasar.controller;

import com.teamzemo.quasar.dto.ApiResponse;
import com.teamzemo.quasar.dto.SpeechMetricsRequest;
import com.teamzemo.quasar.model.Session;
import com.teamzemo.quasar.model.SpeechMetrics;
import com.teamzemo.quasar.repository.SessionRepository;
import com.teamzemo.quasar.repository.SpeechMetricsRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api")
public class SpeechController {

    private static final Logger log = LoggerFactory.getLogger(SpeechController.class);

    private final SpeechMetricsRepository speechMetricsRepository;
    private final SessionRepository sessionRepository;

    public SpeechController(SpeechMetricsRepository speechMetricsRepository,
                             SessionRepository sessionRepository) {
        this.speechMetricsRepository = speechMetricsRepository;
        this.sessionRepository       = sessionRepository;
    }

    /** POST /api/sessions/:sessionId/speech-metrics */
    @PostMapping("/sessions/{sessionId}/speech-metrics")
    public ResponseEntity<ApiResponse<SpeechMetrics>> saveSpeechMetrics(
            @PathVariable String sessionId,
            @RequestBody SpeechMetricsRequest req) {
        try {
            if (req.getTranscript() == null && req.getFillerBuckets() == null) {
                return ResponseEntity.badRequest()
                        .body(ApiResponse.fail("transcript or fillerBuckets required"));
            }

            Session session = sessionRepository.findById(sessionId).orElse(null);
            if (session == null) {
                return ResponseEntity.status(404).body(ApiResponse.fail("Session not found"));
            }

            SpeechMetrics sm = speechMetricsRepository.findBySessionId(sessionId)
                    .orElse(new SpeechMetrics());
            sm.setSessionId(sessionId);
            if (req.getTranscript()     != null) sm.setTranscript(req.getTranscript());
            if (req.getFillerBuckets()  != null) sm.setFillerBuckets(req.getFillerBuckets());
            if (req.getTotalFillers()   != null) sm.setTotalFillers(req.getTotalFillers());
            if (req.getWordsPerMinute() != null) sm.setWordsPerMinute(req.getWordsPerMinute());

            sm = speechMetricsRepository.save(sm);
            return ResponseEntity.ok(ApiResponse.ok("Speech metrics saved", sm));
        } catch (Exception e) {
            log.error("Save speech metrics error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to save speech metrics"));
        }
    }

    /** GET /api/sessions/:sessionId/speech-metrics */
    @GetMapping("/sessions/{sessionId}/speech-metrics")
    public ResponseEntity<ApiResponse<SpeechMetrics>> getSpeechMetrics(@PathVariable String sessionId) {
        try {
            SpeechMetrics sm = speechMetricsRepository.findBySessionId(sessionId).orElse(null);
            if (sm == null) {
                return ResponseEntity.status(404).body(ApiResponse.fail("Speech metrics not found"));
            }
            return ResponseEntity.ok(ApiResponse.ok("Speech metrics retrieved", sm));
        } catch (Exception e) {
            log.error("Get speech metrics error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to get speech metrics"));
        }
    }
}
